core/src/consistency.rs

```diff
@@ -0,0 +1,148 @@
+    pub fn new_terminal(event: ConsistencyEvent) -> Self {
+        Self {
+            event,
+            pending: Vec::new(),
+        }
+    }
+
+    pub fn new_pending(
+        event: ConsistencyEvent,
+        group: ConsistencyGroup,
+        pending_events: Vec<ConsistencyEvent>,
+    ) -> Self {
+        let pending = pending_events
+            .into_iter()
+            .map(|event| PendingConsistency {
+                event,
+                group: group.clone(),
+            })
+            .collect();
+        Self { event, pending }
+    }
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub enum ConsistencyEvent {
+    // CAUSES
+    Publish(Address),                                   // -> Hold
+    AddPendingValidation(Address),                      // -> RemovePendingValidation
+    SignalZomeFunctionCall(snowflake::ProcessUniqueId), // -> ReturnZomeFunctionResult
+
+    // EFFECTS
+    Hold(Address),                                        // <- Publish
+    UpdateEntry(Address, Address),                        // <- Publish, entry_type=Update
+    RemoveEntry(Address, Address),                        // <- Publish, entry_type=Deletion
+    AddLink(Link),                                        // <- Publish, entry_type=LinkAdd
+    RemoveLink(Link),                                     // <- Publish, entry_type=LinkRemove
+    RemovePendingValidation(Address),                     // <- AddPendingValidation
+    ReturnZomeFunctionResult(snowflake::ProcessUniqueId), // <- SignalZomeFunctionCall
+}
+
+#[derive(Clone, Debug, Serialize)]
+struct PendingConsistency {
+    event: ConsistencyEvent,
+    group: ConsistencyGroup,
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub enum ConsistencyGroup {
+    Source,
+    Validators,
+}
+
+impl ConsistencyModel {
+    pub fn process_action(&mut self, action: &Action) -> Option<ConsistencySignal> {
+        use ConsistencyEvent::*;
+        use ConsistencyGroup::*;
+        match action {
+            Action::Commit((entry, crud_link, _)) => {
+                let address = entry.address();
+                let hold = Hold(address.clone());
+                let meta = crud_link.clone().and_then(|crud| match entry {
+                    Entry::App(_, _) => Some(UpdateEntry(crud, address.clone())),
+                    Entry::Deletion(_) => Some(RemoveEntry(crud, address.clone())),
+                    Entry::LinkAdd(link_data) => Some(AddLink(link_data.clone().link)),
+                    Entry::LinkRemove(link_data) => Some(RemoveLink(link_data.clone().link)),
+                    // Question: Why does Entry::LinkAdd take LinkData instead of Link?
+                    _ => None,
+                });
+                let mut pending = vec![hold];
+                meta.map(|m| pending.push(m));
+                let signal =
+                    ConsistencySignal::new_pending(Publish(address.clone()), Validators, pending);
+                self.commit_cache.insert(address, signal);
+                None
```

Comment:
Why is this signal not returned? Does the JS waiter not need to see commits so it would start waiting for the according pending signal? Maybe I'm misunderstanding something here.